contracts/protocol/tokenization/PToken.sol

```diff
@@ -115,25 +119,41 @@ contract PToken is
     ) external virtual override onlyPool {
         _burnScaled(from, receiverOfUnderlying, amount, index);
         if (receiverOfUnderlying != address(this)) {
-            if (timeLockParams.releaseTime != 0) {
-                ITimeLock timeLock = POOL.TIME_LOCK();
-                uint256[] memory amounts = new uint256[](1);
-                amounts[0] = amount;
-
-                timeLock.createAgreement(
-                    DataTypes.AssetType.ERC20,
-                    timeLockParams.actionType,
-                    _underlyingAsset,
-                    amounts,
-                    receiverOfUnderlying,
-                    timeLockParams.releaseTime
-                );
-                receiverOfUnderlying = address(timeLock);
-            }
-            IERC20(_underlyingAsset).safeTransfer(receiverOfUnderlying, amount);
+            _sendToUserOrTimeLock(
+                timeLockParams,
+                POOL.TIME_LOCK(),
+                _underlyingAsset,
+                amount,
+                receiverOfUnderlying
+            );
         }
     }
 
+    function swapAndBurnFrom(
+        address from,
+        address receiverOfUnderlying,
+        uint256 index,
+        DataTypes.TimeLockParams calldata timeLockParams,
+        DataTypes.SwapAdapter calldata swapAdapter,
+        bytes calldata swapPayload,
+        DataTypes.SwapInfo calldata swapInfo
+    ) external virtual override onlyPool returns (uint256 amount) {
+        require(receiverOfUnderlying != address(this));
+        amount = swapAndTransferUnderlyingTo(
+            receiverOfUnderlying,
+            timeLockParams,
+            swapAdapter,
+            swapPayload,
+            swapInfo
+        );
+        _burnScaled(
+            from,
+            receiverOfUnderlying,
+            !swapInfo.exactInput ? amount : swapInfo.maxAmountIn,
```

Comment:
is this impletation more straighttforward?
`swapInfo.exactInput ? swapInfo.maxAmountIn : amount`